processor/groupbytraceprocessor/processor.go

```diff
@@ -0,0 +1,343 @@
+	sp.received <- td
+	return nil
+}
+
+func (sp *groupByTraceProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (sp *groupByTraceProcessor) Start(_ context.Context, _ component.Host) error {
+	// starts the processes listening to the channels
+	go sp.onReleased()
+	go sp.onRemoved()
+	go sp.onTrace()
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (sp *groupByTraceProcessor) Shutdown(ctx context.Context) error {
+	sp.stopped <- ctx
+	return nil
+}
+
+func (sp *groupByTraceProcessor) onTrace() {
+	for {
+		// We have a "main loop" here, to avoid real locks in the hot path. The only lock we have
+		// is in the shutdown path, and the processor's state is checked in the funcs that run
+		// in the future.
+		// The other two buffered channels do modifications to the internal map/ring as part of the same thread
+		// as the main loop, so, there's only one access at any given time. Potentially blocking operations
+		// should be done asynchronously, which isn't 100% the case in this initial implementation.
+		select {
+		case ctx := <-sp.stopped:
+			sp.processShutdown(ctx)
+			return
+		case traceID := <-sp.expired:
+			// in this implementation, this won't return errors
+			sp.processExpired(traceID)
+		case item := <-sp.received:
+			// in this implementation, this won't return errors
+			sp.processBatch(item)
+		}
+	}
+}
+
+func (sp *groupByTraceProcessor) processBatch(batch pdata.Traces) error {
+	for i := 0; i < batch.ResourceSpans().Len(); i++ {
+		if err := sp.processTrace(batch.ResourceSpans().At(i)); err != nil {
+			sp.logger.Info("failed to process trace", zap.Error(err))
+		}
+	}
+
+	return nil
+}
+
+func (sp *groupByTraceProcessor) processTrace(trace pdata.ResourceSpans) error {
+	if trace.IsNil() {
+		// should not happen with the current code
+		return errNilTrace
+	}
+
+	traceID, err := extractTraceID(trace)
+	if err != nil {
+		return fmt.Errorf("couldn't extract trace ID from trace: %w", err)
+	}
+
+	if traceID == nil {
+		return errNilTraceID
+	}
+
+	sTraceID := traceID.String()
+	if _, ok := sp.idToIndex[sTraceID]; ok {
+		// it exists in memory already, just append the spans to the trace in the storage
+		if err := sp.addSpans(traceID, trace); err != nil {
+			return fmt.Errorf("couldn't add spans to existing trace: %w", err)
+		}
+
+		// we are done with this trace, move on
+		return nil
+	}
+
+	// at this point, we determined that we haven't seen the trace yet, so, record the
+	// traceID in the map and the spans to the storage
+
+	// calculates the item in the ring that we'll store the trace
+	sp.index = (sp.index + 1) % sp.config.NumTraces
+
+	// see if the ring has an item already
+	idAtRingItem := sp.ids[sp.index]
+	if idAtRingItem != nil {
+		// delete from the map and erase its memory entry
```

Comment:
Why not to do it together with the storage removal in processing items from the `removed` channel?
Optimization for potentially slow storage?

Reply:
Yes, storage is potentially blocking (just not in the current implementation).